[ty] Recognize overlapping NewTypes and their underlying values (#27522)
astral-sh/ruff: crates/ty_python_semantic/resources/mdtest/narrow/conditionals/is.md

    reveal_type(y)  # revealed: A | None
```

Identity also transfers facts about the shared object, such as whether a string is truthy.

```py
def truthy_string(value: object, text: str) -> None:
    if text:
        if value is text:
            reveal_type(value)  # revealed: str & ~AlwaysFalsy
```

## `is` with invariant generic types

A `list[int]` guarantees that values read from the list are integers. That guarantee must hold for
every reference to the same mutable list: if another reference could treat it as `list[str]`, it
could append a string that the first reference would then incorrectly read as an integer. An

> identity comparison can therefore transfer the invariant type argument.

Reviewer: Woah. Cool! This makes sense to me. Does any other type checker do this?

Author: Pyrefly does (and we already do on main -- this is just clarifying the rationale for why we can). In this example ty and pyrefly infer `list[int]`, pyright infers `object`, and mypy infers `list[Any]`:

```py
def f(value: object, items: list[int]) -> None:
    if value is items:
        reveal_type(value)
```

```py
def generic_type(value: object, items: list[int]) -> None:
    if value is items:
        reveal_type(value)  # revealed: list[int]
```

Incompatible invariant specializations cannot describe the same object in soundly typed code.

```py
def incompatible_generic_types(integers: list[int], strings: list[str]) -> None:

>     reveal_type(integers is strings)  # revealed: Literal[False]

Reviewer: ```suggestion
    reveal_type(integers is strings)  # revealed: Literal[False]
    if integers is strings:
        reveal_type(integers)  # revealed: Never
        reveal_type(strings)  # revealed: Never
```

I haven't checked that these pass, but I think it's the logical implication of your prose above?

```

## `is` with covariant generic types

Covariant specializations can describe the same object: an empty tuple belongs to both
`tuple[int, ...]` and `tuple[str, ...]`. Identity therefore remains possible and preserves both sets
of type arguments.

```py
def covariant_generic_type(value: object, items: tuple[int, ...]) -> None:
    if value is items:
        reveal_type(value)  # revealed: tuple[int, ...]

def overlapping_generic_types(integers: tuple[int, ...], strings: tuple[str, ...]) -> None:
    reveal_type(integers is strings)  # revealed: bool
    if integers is strings:
        # TODO: Ideally, this intersection would simplify to tuple[()].

>         reveal_type(integers)  # revealed: tuple[int, ...] & tuple[str, ...]

Reviewer: This should also narrow the r.h.s. too, I think, could add an assertion for that as well

```

## `is` with a `NewType`

A `NewType` constructor returns its argument unchanged, so its tag belongs to one static view rather
than the shared object. Identity can establish the underlying type without transferring that tag.

```py
from typing import NewType

UserId = NewType("UserId", int)

def discard_newtype_tag(value: object, user_id: UserId) -> None:
    if value is user_id:

>         reveal_type(value)  # revealed: int

Reviewer: Nice!

        reveal_type(value)  # revealed: int | (None & ~T@f) | (EllipsisType & ~T@f)
```

## `is` with a negated `NewType`

Excluding a `NewType` removes its invisible tag, not the runtime objects accepted by its

> constructor. An identity comparison must preserve that negation without making a reachable branch

Reviewer: ```suggestion
constructor. An identity comparison preserves that negation without making a reachable branch
```

```

A type variable can hide the same static negation in its upper bound. The reachable branch must
preserve that type variable.

```py
ExcludedBound = TypeVar("ExcludedBound", bound=Intersection[int, Not[UserId]])

def excluded_newtype_in_bound(
    value: ExcludedBound,
    other: UserId,
    without_one: Intersection[ExcludedBound, Not[Literal[1]]],
) -> None:
    if value is other:
        reveal_type(value)  # revealed: ExcludedBound@excluded_newtype_in_bound

    if without_one is other:
        reveal_type(without_one)  # revealed: ExcludedBound@excluded_newtype_in_bound & ~Literal[1]
```

The same runtime overlap remains reachable when both operands are unions, while genuinely
incompatible alternatives are removed.

```py
def excluded_newtype_in_unions(
    value: Intersection[int, Not[UserId]] | None,
    other: UserId | bytes,
) -> None:
    if value is other:
        reveal_type(value)  # revealed: int & ~UserId
        reveal_type(other)  # revealed: UserId
```

Unlike a negated `NewType`, a negated runtime class genuinely rules out identity with its instances.

```py
def excluded_runtime_class(not_int: Not[int], other: UserId) -> None:
    if not_int is other:

>         reveal_type(not_int)  # revealed: Never

Reviewer: ```suggestion
        reveal_type(not_int)  # revealed: Never
        reveal_type(other)  # revealed: Never
```

```

## `is` does not transfer `LiteralString`

`LiteralString` records string provenance, not a guarantee shared by every view of the same string
object. An identity comparison can establish that another value is a string, but cannot transfer
that provenance. A concrete string literal still identifies its runtime value.

```py
from typing import Any, Literal, TypeVar
from typing_extensions import LiteralString, TypeIs
from ty_extensions import Intersection, Not

def literal_string(value: object, text: LiteralString) -> None:
    if value is text:
        reveal_type(value)  # revealed: str
        reveal_type(text)  # revealed: LiteralString

def concrete_literal(value: object, text: Literal["hello"]) -> None:
    if value is text:
        reveal_type(value)  # revealed: Literal["hello"]
```

Negating `LiteralString` likewise does not rule out identity with a concrete string.

```py
def excluded_literal_string(
    value: Not[LiteralString],

>     text: Intersection[str, Not[LiteralString]],

Reviewer: You can write this as `str & ~LiteralString` if you make the snippet a pyi snippet (we allow the experimental syntax in stub files now). I think there's a few opportunities elsewhere in this PR where you could do the same, too

Author: I don't really like using pyi for that reason, when the test has actual function-body code in it that would never appear in a real pyi file. I'd prefer tests being more "realistic" over the ability to use the nice intersection syntax.

Reviewer: fair

    other: Literal["hello"],
) -> None:
    if value is "hello":
        reveal_type(value)  # revealed: str & ~LiteralString

    if text is "hello":
        reveal_type(text)  # revealed: str & ~LiteralString

    if value is other:
        reveal_type(value)  # revealed: str & ~LiteralString
        reveal_type(other)  # revealed: Literal["hello"]
```

A gradual string literal can arise naturally by narrowing an `Any` value. Comparing it with a string
whose `LiteralString` provenance was negated must not make the reachable branch disappear.

```py
def is_literal_string(value: str) -> TypeIs[LiteralString]:
    return False

def gradual_literal_from_narrowing(value: str, untyped: Any) -> None:
    if not is_literal_string(value):
        if untyped is "hello":

>             reveal_type(untyped)  # revealed: Any & Literal["hello"]

Reviewer: Does this not challenge the idea that all literal string types are subtypes of `LiteralString`, if we allow a value of type `~LiteralString` to be narrowed to a literal string type?

Reviewer: I think the logic of this PR (which I think is... correct?) implies that `Literal` string types are not subtypes of `LiteralString`: `Literal["hello"]` is actually a distinct type to `LiteralString & Literal["hello"]`, and the latter is unsimplifiable.

Author: This test is confusing as written. But this PR does not "allow a value of type `~LiteralString` to be narrowed to a literal string type", nor does that occur in this test. The test independently constructs `value` of type `str & ~LiteralString` and `untyped` of type `Any & Literal["hello"]`, via uncontroversial narrowings, and then it asserts that `if value is untyped` is not considered unreachable, and does not further narrow either type.

That said, I think you're right that this PR (and this test) exposes some inherent inconsistency in the semantics of `LiteralString` and string `Literal` types. The core problem is that these three things, which independently seem intuitive/desirable, cannot all simultaneously be true:

1. Every runtime string object with value `"hello"` inhabits the type `Literal["hello"]`.
2. `Literal["hello"]` is a subtype of `LiteralString`.
3. There exist string objects which do not inhabit `LiteralString`.

Since this gets potentially pretty hairy, I'm going to push an update that avoids touching literal-string narrowing logic as much as possible in this PR, and I'll defer it to a separate issue.

variable without making the concrete-string comparison unreachable.

```py
NonLiteralStringT = TypeVar("NonLiteralStringT", bound=Intersection[str, Not[LiteralString]])

def excluded_literal_string_in_bound(value: NonLiteralStringT, other: Literal["hello"]) -> None:
    if value is other:
        reveal_type(value)  # revealed: NonLiteralStringT@excluded_literal_string_in_bound
```

The same negation remains compatible with a concrete string when both operands contain additional
alternatives. Unrelated alternatives are still removed, while compatible alternatives are retained.

```py
def excluded_literal_string_in_union(
    value: Intersection[str, Not[LiteralString]] | int,
    other: Literal["hello"] | bytes,
) -> None:
    if value is other:
        reveal_type(value)  # revealed: str & ~LiteralString
        reveal_type(other)  # revealed: Literal["hello"]

def excluded_literal_string_with_shared_alternative(
    value: Intersection[str, Not[LiteralString]] | bytes,
    other: Literal["hello"] | bytes,
) -> None:
    if value is other:
        reveal_type(value)  # revealed: bytes | (str & ~LiteralString)
        reveal_type(other)  # revealed: Literal["hello"] | bytes
```

> In contrast, excluding a concrete literal genuinely rules out identity with that value.

Reviewer: ```suggestion
In contrast, excluding a concrete literal genuinely rules out identity with string literals that have that value.
```

BoundedT = TypeVar("BoundedT", bound=FooNewType1)
BoundedU = TypeVar("BoundedU", bound=FooNewType2)

> def bounded_typevars(left: BoundedT, right: BoundedU) -> None:

Reviewer: Why is the return type of this function changing? This was added in Charlie's PR specifically as a TODO test for an issue I pointed out with his PR, to demonstrate that "narrowing away" typevars could lead to real (albeit unlikely) false-positive diagnostics

Author: I think the TODO in this test was wrong, and the return error that it called a false positive was in fact a true positive. The problem is that it's unsound to have `left is right` narrow either `left` or `right` to an intersection of the two typevars, because typevars can be solved to a type with static tags (a NewType or a generic type) that isn't fully validated by runtime `is` comparison. Here's a fuller example of the potential unsoundness:

```py
from typing import NewType, TypeVar

UserId = NewType("UserId", int)
OrderId = NewType("OrderId", int)

T = TypeVar("T")
U = TypeVar("U")

def choose_right(left: T, right: U) -> U:
    if left is right:
        return left  # Must remain an error: T is not necessarily U.
    return right

raw = 123
user = UserId(raw)
order = OrderId(raw)

assert user is order  # NewType constructors return their argument unchanged.

# If we allowed `choose_right` to return `user` based on `user is order`, that would be unsound:
result: OrderId = choose_right(user, order)
```

We could decide that this is an edge case unsoundness we don't care about, if narrowing typevars by identity turns out to be important? But I'd rather evaluate that in a separate PR. In the meantime, I don't think we should describe it here as a false positive when it is an intentional soundness choice (or at least, became one in this PR!)

We could restore the return type and assert the return error as a true positive, but tbh that seems redundant with the existing assertions on the type of `left` and `right`.

Reviewer: ah, I see. That makes sense to me -- thanks for the explanation!

Author: Ok so on second look, that rationale didn't quite look convincing to me -- in that particular example given that `left` and `right` must be the _same runtime object_, it can't possibly cause unsoundness to return one vs the other. So I think in that sense we could consider the return error here a false positive and aim to fix it in the future.

The tricky part is that we can't soundly achieve that goal via just inferring an intersection of the two typevars, because there are cases where that intersection type could escape, and that would be unsound. Here's such a case:

```py
from typing import Generic, NewType, TypeVar

UserId = NewType("UserId", int)
OrderId = NewType("OrderId", int)

T = TypeVar("T")
U = TypeVar("U")

class Common(Generic[T, U]):
    def __init__(self, left: T, right: U) -> None:
        if left is right:
            self.value = left
        else:
            self.value = None

raw = 123
result = Common(UserId(raw), OrderId(raw))

if result.value is not None:
    result.value.does_not_exist()
```

We would end up inferring `result.value` as `(UserId & OrderId) | None`, which would simplify to `None` and we'd wrongly consider the final branch unreachable.

So if we want to address the false positive that was originally recorded in this example, we would have to do it via tracking identity relationships "out of band" in some way, we shouldn't do it by inferring `T & U`.

I can add a comment here to better explain this.

    reveal_type(left is right)  # revealed: bool
    if left is right:
        reveal_type(left)  # revealed: BoundedT@bounded_typevars
        reveal_type(right)  # revealed: BoundedU@bounded_typevars

ConstrainedT = TypeVar("ConstrainedT", FooNewType1, FooNewType2)
ConstrainedU = TypeVar("ConstrainedU", FooNewType3, FooNewType4)

> def constrained_typevars(left: ConstrainedT, right: ConstrainedU) -> None:

Reviewer: Same here, I think the test was better with the previous return type, because it demonstrates that if we do the wrong thing then there's a false-positive `invalid-return-type` diagnostic on the user's code

Author: See above.
